compiler/rustc_builtin_macros/src/autodiff.rs

```diff
@@ -0,0 +1,707 @@
+#![allow(unused_imports)]
+use std::str::FromStr;
+use std::string::String;
+
+use rustc_ast::expand::autodiff_attrs::{
+    is_fwd, is_rev, valid_input_activity, valid_ty_for_activity, AutoDiffAttrs, DiffActivity,
+    DiffMode,
+};
+use rustc_ast::ptr::P;
+use rustc_ast::token::{Token, TokenKind};
+use rustc_ast::tokenstream::*;
+use rustc_ast::visit::AssocCtxt::*;
+use rustc_ast::Safety::Unsafe;
+use rustc_ast::{
+    self as ast, AssocItemKind, BindingMode, ByRef, Fn, FnHeader, FnRetTy, FnSig, Generics,
+    ItemKind, MetaItemKind, NestedMetaItem, PatKind, Stmt, StmtKind, TyKind,
+};
+use rustc_expand::base::{Annotatable, ExtCtxt};
+use rustc_span::symbol::{kw, sym, Ident};
+use rustc_span::{Span, Symbol};
+use thin_vec::{thin_vec, ThinVec};
+use tracing::trace;
+
+use crate::errors;
+
+#[cfg(not(not(llvm_enzyme)))]
+pub fn expand(
+    ecx: &mut ExtCtxt<'_>,
+    _expand_span: Span,
+    meta_item: &ast::MetaItem,
+    item: Annotatable,
+) -> Vec<Annotatable> {
+    ecx.sess.dcx().emit_err(errors::AutoDiffSupportNotBuild { span: meta_item.span });
+    return vec![item];
+}
+
+#[cfg(not(llvm_enzyme))]
+fn first_ident(x: &NestedMetaItem) -> rustc_span::symbol::Ident {
+    let segments = &x.meta_item().unwrap().path.segments;
+    assert!(segments.len() == 1);
+    segments[0].ident
+}
+
+#[cfg(not(llvm_enzyme))]
+fn name(x: &NestedMetaItem) -> String {
+    first_ident(x).name.to_string()
+}
+
+#[cfg(not(llvm_enzyme))]
+pub fn from_ast(
+    ecx: &mut ExtCtxt<'_>,
+    meta_item: &ThinVec<NestedMetaItem>,
+    has_ret: bool,
+) -> AutoDiffAttrs {
+    let mode = name(&meta_item[1]);
+    let mode = match DiffMode::from_str(&mode) {
+        Ok(x) => x,
+        Err(_) => {
+            ecx.sess
+                .dcx()
+                .emit_err(errors::AutoDiffInvalidMode { span: meta_item[1].span(), mode });
+            return AutoDiffAttrs::inactive();
+        }
+    };
+    let mut activities: Vec<DiffActivity> = vec![];
+    for x in &meta_item[2..] {
+        let activity_str = name(&x);
+        let res = DiffActivity::from_str(&activity_str);
+        match res {
+            Ok(x) => activities.push(x),
+            Err(_) => {
+                ecx.sess.dcx().emit_err(errors::AutoDiffUnknownActivity {
+                    span: x.span(),
+                    act: activity_str,
+                });
+                return AutoDiffAttrs::inactive();
+            }
+        };
+    }
+
+    // If a return type exist, we need to split the last activity,
+    // otherwise we return None as placeholder.
+    let (ret_activity, input_activity) = if has_ret {
+        activities.split_last().unwrap()
```

Comment:
We count the number of input+output args, and assert that we got one activity for each, or we won't reach this code path. But just in case I added a match and a note.